pkg/virt-launcher/virtwrap/cmd-server/client/client.go

```diff
@@ -0,0 +1,238 @@
+)
+
+type Reply struct {
+	Success        bool
+	Message        string
+	DomainListJSON string
+}
+
+type Args struct {
+	// used for domain management
+	VMJSON          string
+	K8SecretMapJSON string
+
+	// used for syncing secrets
+	SecretUsageType string
+	SecretUsageID   string
+	SecretValue     string
+}
+
+type LauncherClient interface {
+	StartVirtualMachine(vm *v1.VirtualMachine, secrets map[string]*k8sv1.Secret) error
+	ShutdownVirtualMachine(vm *v1.VirtualMachine) error
+	KillVirtualMachine(vm *v1.VirtualMachine) error
+	SyncSecret(vm *v1.VirtualMachine, usageType string, usageID string, secretValue string) error
+	ListDomains() ([]*api.Domain, error)
+	Ping() error
+	Close()
+}
+
+type VirtLauncherClient struct {
+	client *rpc.Client
+}
+
+func ListAllSockets(baseDir string) ([]string, error) {
+	var socketFiles []string
+
+	fileDir := filepath.Join(baseDir, "sockets")
+	exists, err := diskutils.FileExists(fileDir)
+	if err != nil {
+		return nil, err
+	}
+
+	if exists == false {
+		return socketFiles, nil
+	}
+
+	files, err := ioutil.ReadDir(fileDir)
+	if err != nil {
+		return nil, err
+	}
+	for _, file := range files {
+		socketFiles = append(socketFiles, filepath.Join(fileDir, file.Name()))
+	}
+	return socketFiles, nil
+}
+
+func SocketsDirectory(baseDir string) string {
+	return filepath.Join(baseDir, "sockets")
+}
+
+func SocketFromNamespaceName(baseDir string, namespace string, name string) string {
+	sockFile := namespace + "_" + name + "_sock"
+	return filepath.Join(SocketsDirectory(baseDir), sockFile)
+}
+
+func GetClient(socketPath string) (LauncherClient, error) {
+	conn, err := rpc.Dial("unix", socketPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &VirtLauncherClient{client: conn}, nil
+}
+
+func (c *VirtLauncherClient) Close() {
+	c.client.Close()
+}
+
+func (c *VirtLauncherClient) genericSendCmd(args *Args, cmd string) (*Reply, error) {
+	reply := &Reply{}
+
+	err := c.client.Call(cmd, args, reply)
+	if IsDisconnected(err) {
+		return reply, err
+	} else if err != nil {
+		msg := fmt.Sprintf("unknown error encountered sending command %s: %s", cmd, err.Error())
+		return reply, fmt.Errorf(msg)
+	} else if reply.Success != true {
+		msg := fmt.Sprintf("server error. command %s failed: %s", cmd, reply.Message)
+		return reply, fmt.Errorf(msg)
+	}
+	return reply, nil
+}
+
+func (c *VirtLauncherClient) ShutdownVirtualMachine(vm *v1.VirtualMachine) error {
+	cmd := "Launcher.Shutdown"
+
+	vmJSON, err := json.Marshal(*vm)
+	if err != nil {
+		return err
+	}
+	args := &Args{
+		VMJSON: string(vmJSON),
+	}
+	_, err = c.genericSendCmd(args, cmd)
+
+	return err
+}
+
+func (c *VirtLauncherClient) KillVirtualMachine(vm *v1.VirtualMachine) error {
+	cmd := "Launcher.Kill"
+
+	vmJSON, err := json.Marshal(*vm)
+	if err != nil {
+		return err
+	}
+	args := &Args{
+		VMJSON: string(vmJSON),
+	}
+	_, err = c.genericSendCmd(args, cmd)
+
+	return err
+}
+
+func (c *VirtLauncherClient) ListDomains() ([]*api.Domain, error) {
```

Comment:
Shouldn't that be `ListDomain`?

Reply:
This translates to ListAllDomains on the backend.